### DISCUSSION TITLE
How to inject content script into the iframe on the page？

### DISCUSSION THREAD

**Discussion by asasas234** [wxt-dev/wxt#1183] [Q&A]
I tried using the following code, setting allFrames to true did not work.
```
export default defineContentScript({
    matches: ["*://*/*"], // 必须要配置匹配youtube所有页面，如果只配置youtube视频页，那么从其他页面先进入视频页，不会触发main方法的执行
    excludeMatches: ["*://repository.modelio.org/*"],
    runAt: "document_end",
    allFrames: true,
    async main(ctx){
       
    }

})
```

**Comment by asasas234** [comment]
It should be noted that I found that the iframe was not created during the document_end execution phase, which may be the reason why allFrames did not take effect. In this case, how can we inject content scripts? @aklinker1

**Reply by aiktb** [reply]
> It should be noted that I found that the iframe was not created during the document_end execution phase, which may be the reason why allFrames did not take effect. In this case, how can we inject content scripts? @aklinker1

https://developer.chrome.com/docs/extensions/develop/concepts/content-scripts#:~:text=script%20is%20run.-,document_end,-string

| Name              | Type   | Description                                                                                                                                                                                                                                                                                                                                                                                   |
|-------------------|--------|-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| document_idle     | string | **Preferred.** Use "document_idle" whenever possible. The browser chooses a time to inject scripts between "document_end" and immediately after the `window.onload` event fires. The exact moment of injection depends on how complex the document is and how long it is taking to load, and is optimized for page load speed. Content scripts running at "document_idle" don't need to listen for the `window.onload` event, they are guaranteed to run after the DOM is complete. If a script definitely needs to run after `window.onload`, the extension can check if `onload` has already fired by using the `document.readyState` property. |
| document_end      | string | Scripts are injected immediately after the DOM is complete, **_but before subresources like images and frames have loaded._** |

Try `document_idle`

**Comment by asasas234** [comment]
```
export default defineContentScript({
    matches: ["*://*/*"], 
    excludeMatches: ["*://repository.modelio.org/*"],
    runAt: "document_idle",
    allFrames: true,
    async main(ctx){
        // content-script.js
        if (window.self !== window.top) {
            console.log("Injected into iframe");
        } else {
            console.log("Injected into top window");
        }
    }

})
```
I tested it and it didn't work. It didn't execute inside the if statement. @aiktb

**Reply by aiktb** [reply]
idk🤷, `allFrames: true` works fine in my project, need a reproducible.

**Reply by MentalGear** [reply]
Thanks, this works with:
`js: ["content-scripts/content.js"],`

Can you recommended a way to handle iframes that are loaded dynamically, e.g. comments widgets on the bottom of a page. How to ensure the iframe content is injected there as well?

**Comment by asasas234** [comment]
@aiktb https://www.gamedev.tv/dashboard/courses/63  You can try this webpage. This course is one of the free courses on this website, you just need to register an account and go through a payment process (but no actual payment is required).

**Comment by asasas234** [comment]
@aiktb Or you can give me an email address, and I can send you my account for testing.

**Comment by asasas234** [comment]
@aiktb Thank you, I found that it has taken effect, although the breakpoint did not hit, but I found that the console is outputting normally.
